### PR TITLE
receipt with current standing

### PR DESCRIPTION
**Pull request by otahirs**
#361
first version of more detailed receipt 
still missing relay information
needs to be tested on thermoprinter

Capture

### CODE REVIEW

**Comment by lukaskett** [comment]
It looks fine, I see all required data fields, mainly loss to leading runner, well done. My thoughts:

1. Change date format from computer to human - 2022-01-22 -> 22.01.2022
2. I miss SI card number, useful when the SI card is not assigned to the runner. Could be in the header under the name.
3. Hide 0 elevation if possible
4. Bold highlight the best split times, both on leg and intermediate like the final time is
5. Align times in the columns according to the colons, especially the second one
6. Add plus sign to the loss to leading runner, the sign could be minus in case of the new best time, no idea how it works right now
7. Can we add battery status? #363

**Comment by arnost00** [comment]
> It looks fine, I see all required data fields, mainly loss to leading runner, well done. My thoughts:
> 
>     1. Change date format from computer to human - 2022-01-22 -> 22.01.2022

22.01.2022 is Czech format. I thing the best way is use date format according to language settings of system / application. 
QLocale::system().toString(date, QLocale::ShortFormat);

**Comment by kamilpipek** [comment]
Concerning the SIAC battery voltage there is good experience with scenario described. Only one irregularity found for SIAC cards FW revision 3.2 or older (it returns the value 0xEE which means 23.31V).

**Comment by paukert** [comment]
I noticed one more thing - I would remove space between + sign and loss on split time

File: quickevent/app/quickevent/plugins/Receipts/qml/receipts/Complete (beta).qml [hunk 1]
```diff
@@ -0,0 +1,402 @@
+import QtQml 2.0
```
Comment (paukert) [inline quickevent/app/quickevent/plugins/Receipts/qml/receipts/Complete (beta).qml:1]: Can you fix indent, please? There are spaces as well as tabs.

**Comment by otahirs** [comment]
>       22.01.2022 is Czech format. I thing the best way is use date format according to language settings of system / application. QLocale::system().toString(date, QLocale::ShortFormat);

I often use eng locale on my machine and i dont want to see 01/22/2022 on the receipts
>     4. Bold highlight the best split times, both on leg and intermediate like the final time is
>     5. Align times in the columns according to the colons, especially the second one
done

>     6. Add plus sign to the loss to leading runner, the sign could be minus in case of the new best time, no idea how it works right now
new best time shows as `loss to leading runner = 0:00`
>     7. Can we add battery status? [Print SIAC battery status on receipt #363](https://github.com/Quick-Event/quickbox/issues/363)
someone can, but i am not doing it now

>  I would remove space between + sign and loss on split time

the space is there for a purpose
image

---

image

image

**Comment by otahirs** [comment]
TODO:
- [ ] test and tweak for thermal printers
- [ ] add relay information 
> V případě štafet by bylo dobré na lísteček doplnit číslo úseku, startovní číslo, čas na úseku, čas po n úsecích a jestli je ražení OK(např. 2.úsek může být OK, ale 1.úsek není, takže celkem to dá DISK).

**Comment by paukert** [comment]
> the space is there for a purpose
image

I thought that + sign would be just time prefix and would not be aligned if loss is greater than 10 minutes but it does not really matter

**Comment by otahirs** [comment]
image

---
image

---
tweaked for thermal printers, still not tested on a real one

**Comment by otahirs** [comment]
doplněno jméno štafety (zkratka klubu, číslo),  číslo štafety, úseku
čas po n úsecích mi přijde zbytečný
pořadí po N úsecích je další SQL rabbit hole do kterého se může pustit někdo další

zvláště mezičasy budou u štafet asi velmi orientační, a při jiném počtu kontrol se to může předpokládám trochu rozpadnout, takze zde trochu best effort. u normalnich zavodu by vse melo byt ok

**Comment by otahirs** [comment]
272665531_397425612141260_7423484575693257066_n
271514378_313941800672672_1390812723511734997_n

ukazky termotisku od @lukaskett , troje zvojene *underline* u druheho obrazku jsem odstranil
duvod posunuti podtrhnuti u hlavicky je mi zahadou

File: quickevent/app/quickevent/plugins/Receipts/src/receiptsplugin.cpp [hunk 2]
```diff
@@ -184,108 +184,163 @@ QVariantMap ReceiptsPlugin::receiptTablesData(int card_id)
 	bool is_card_lent = getPlugin<CardReaderPlugin>()->isCardLent(card_id, read_card.finishTime(), run_id);
```
Comment (paukert) [inline quickevent/app/quickevent/plugins/Receipts/src/receiptsplugin.cpp:184]: It is not directly related to this PR but IMO this does not work properly. First argument of function isCardLent should SI card number, not id from cards table.

Comment (otahirs) [reply quickevent/app/quickevent/plugins/Receipts/src/receiptsplugin.cpp:184]: afaik `isCardLent` takes id, not card number, `card_id` is an abbreviation of `id` form table `cards`
 it surely does not help that si card number is named `siId` in the db schema (the table `cards` contains both `id` and `siId` columns)

Comment (paukert) [reply quickevent/app/quickevent/plugins/Receipts/src/receiptsplugin.cpp:184]: There is a query in `isCardLent` where is surely card number expected:
`q.exec("SELECT siId FROM lentcards WHERE NOT ignored AND siid=" QF_IARG(si_id) );`

and also `findRunId` is called which also takes card number

File: quickevent/app/quickevent/plugins/Receipts/src/receiptsplugin.cpp [hunk 3]
```diff
@@ -184,108 +184,163 @@ QVariantMap ReceiptsPlugin::receiptTablesData(int card_id)
...
-				}
-				else {
-					int class_id = model.value(0, "competitors.classId").toInt();
-					qb_minlaps.select("runlaps.position, MIN(runlaps.lapTimeMs) AS minLapTimeMs")
-							.from("competitors")
-							.joinRestricted("competitors.id", "runs.competitorId", "runs.stageId=" QF_IARG(current_stage_id) " AND competitors.classId=" QF_IARG(class_id), qf::core::sql::QueryBuilder::INNER_JOIN)
-							.joinRestricted("runs.id", "runlaps.runId", "runlaps.position > 0 AND runlaps.lapTimeMs > 0", qf::core::sql::QueryBuilder::INNER_JOIN)
-							.groupBy("runlaps.position")
-							.orderBy("runlaps.position");
+			qf::core::sql::Query run;
```
Comment (fvacek) [inline quickevent/app/quickevent/plugins/Receipts/src/receiptsplugin.cpp:220]: `qf::core::sql::Query run_laps;` prosim

**Comment by fvacek** [comment]
tohle review necham na @paukert

**Comment by paukert** [comment]
Pokusil jsem se to nějak projít a na nic zásadního jsem nenarazil..

Musím se ale přiznat, že s Qt/Qml/C++ pracuji minimálně (v současné době pouze ve spojení s QE) a code base QE také do detailů neznám, takže nevím, jestli jsem úplně ten správný člověk na review :D

**Comment by paukert** [comment]
Možná rovnou pořešit i https://github.com/Quick-Event/quickbox/issues/671

**Comment by fvacek** [comment]
> Pokusil jsem se to nějak projít a na nic zásadního jsem nenarazil..
> 
> Musím se ale přiznat, že s Qt/Qml/C++ pracuji minimálně (v současné době pouze ve spojení s QE) a code base QE také do detailů neznám, takže nevím, jestli jsem úplně ten správný člověk na review :D

V tomhle je na @otahirs už docela spolehnutí (myslím c++ a qml) :) Stačí, když dohledné na to, že je usteizek po formální a obsahové stránce krok kupředu.

**Comment by paukert** [comment]
> > Pokusil jsem se to nějak projít a na nic zásadního jsem nenarazil..
> > Musím se ale přiznat, že s Qt/Qml/C++ pracuji minimálně (v současné době pouze ve spojení s QE) a code base QE také do detailů neznám, takže nevím, jestli jsem úplně ten správný člověk na review :D
> 
> V tomhle je na @otahirs už docela spolehnutí (myslím c++ a qml) :) Stačí, když dohledné na to, že je usteizek po formální a obsahové stránce krok kupředu.

Po obsahové (formální) stránce je IMO nový lísteček určitě krok kupředu. Komu by se náhodou nelíbil, tak stále má možnost využívat "starou" verzi :)

**Comment by otahirs** [comment]
ještě můžeme přijít s lepším názvem, zatím tam je `Complete (beta)`

**Comment by fvacek** [comment]
> ještě můžeme přijít s lepším názvem, zatím tam je `Complete (beta)`

ja nemam moc rad mezery v nazvech souboru, tak bych se primlouval za `Complete-beta`

**Comment by paukert** [comment]
> > ještě můžeme přijít s lepším názvem, zatím tam je `Complete (beta)`
> 
> ja nemam moc rad mezery v nazvech souboru, tak bych se primlouval za `Complete-beta`

Oni ty názvy lístečků nejsou obecně úplně vypovídající, protože člověku těžko řekne něco více název Classic a Complete. Complete ve mně evokuje, že bude obsahovat všechny (možné) údaje, ale ve skutečnosti neobsahuje nějaké věci, co má zase Classic.

Aby to bylo více vypovídající, jak by název obou lístečků musel být delší.. U classicu by asi v názvu mohlo být `laps-pace` (obsahuje "tempo" mezi kontrolami), u complete asi `laps-standings` (obsahuje pořadí na kontrolách + průběžné pořadí na kontrolách).

**Comment by otahirs** [comment]
`laps-pace` a `laps-standings` zase nikomu krome nas nic nerekne.
mozna bych zustal u neceho jako `Classic` pro konzervy, aby se jim nic nezmenilo, a novy bych oznacil jako `Default`

**Comment by arnost00** [comment]
Nejsme schopni k nazvu listecku dodat nejaky komentar ?
V nem by mohlo byt vysvetleno co ktery obsahuje (tempo / poradi na kontrolach atp.)
Pak by to snad mohlo bych pochopitelne i pro nove uzivatele QE.

**Comment by fvacek** [comment]
jak to tady vypada, neni to uz nahodou hotovy?

**Comment by otahirs** [comment]
Je. Ještě teda prejmejuju `Complete (beta)` na `Default`

**Comment by arnost00** [comment]
Na teto verzi **+ db postgres = nefunguji listecky s mezicasy**.
Nahled listecku shodi program, a automaticky tisk listecku pise pouze chyby do logu programu. 
Zavod 2.4. byl nutny prechod na starsi verzi.

Vyber listecku mezi Default a Classic - vubec jsem si nebyl schopen rano z hlavy vybavit ktery je ktery. Asi bych si musel precist tuto komunikaci nebo issue. Nebo metoda nahodneho vyberu.

**Comment by arnost00** [comment]
Po zavode testovano znovu - na sqlite to jede na postgresu ne. Testovano na pg serveru v13

**Comment by arnost00** [comment]
2022-04-02T16:15:14.097	Error		qf::core::Exception: SELECT COUNT(*) as currStanding FROM (SELECT runs.id, SUM(runlaps.lapTimeMs) as currTime FROM competitors INNER JOIN runs ON runs.competitorId=competitors.id INNER JOIN runlaps ON runlaps.runId=runs.id WHERE (runs.stageId=1 AND competitors.classId=155618) AND (runlaps.position > 0 AND runlaps.lapTimeMs > 0 AND runlaps.position<=1 AND NOT runs.disqualified AND runs.isRunning AND NOT runs.misPunch) GROUP BY runs.id) WHERE currTime < 71000\nERROR:  subquery in FROM must have an alias\nLINE 1: SELECT COUNT(*) as currStanding FROM (SELECT runs.id, SUM(ru...\n                                             ^\nHINT:  For example, FROM (SELECT ...) [AS] foo.\n(42601) QPSQL: Nepodařilo se vytvořit žádný dotaz	receiptsplugin.cpp	479

**Comment by fvacek** [comment]
fixed in 2.5.6

**Pull request merged**